### PR TITLE
[Basic] Add TemporaryDirectory class.

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
This class is a wrapper over POSIX's mkdtemp() function useful to create
disposable directories.

### CODE REVIEW

File: Sources/Basic/TemporaryFile.swift [hunk 1]
```diff
@@ -31,6 +32,27 @@ private extension TempFileError {
...
 }
 
+/// Determines the directory in which the temporary file should be created. Also makes
+/// sure the returning path has a trailing forward slash.
+///
+/// - Parameters:
+///     - dir: If present this will be the temporary directory.
+///
+/// - Returns: Path to directory in which temporary file should be created.
+private func determineTempDirectory(_ dir: String? = nil) -> String {
```
Comment (ddunbar) [inline Sources/Basic/TemporaryFile.swift]: Can we share this with the TemporaryFile implementation?

File: Sources/Basic/TemporaryFile.swift [hunk 2]
```diff
@@ -86,31 +108,107 @@ public final class TemporaryFile {
...
+}
+
+extension TemporaryFile: CustomStringConvertible {
+    public var description: String {
+        return "<TemporaryFile: \(path)>"
+    }
+}
+
+/// FIXME: This isn't right place to declare this, probably POSIX or merge with FSProxyError?
+/// Contains the error which can be thrown while creating a directory using POSIX's mkdir.
```
Comment (ddunbar) [inline Sources/Basic/TemporaryFile.swift]: Yeah... I'm not sure how to handle this yet I am still gaining experience with using Swift Errors...

Nit: I don't think FIXME's should be doc comments, usually

File: Sources/Basic/TemporaryFile.swift [hunk 3]
```diff
@@ -86,31 +108,107 @@ public final class TemporaryFile {
...
+    }
+}
+
+/// A class to create disposable directories using POSIX's mkdtemp() method.
+public final class TemporaryDirectory {
+    /// If specified during init, the temporary directory name begins with this prefix.
+    let prefix: String
+
+    /// The directory in which the temporary directory is created.
+    public let dir: String
```
Comment (ddunbar) [inline Sources/Basic/TemporaryFile.swift]: This is always going to be `path.parentDirectory`, right? If so I'm not sure it is worth having both, a simpler API seems nicer.

File: Sources/Basic/TemporaryFile.swift [hunk 4]
```diff
@@ -86,31 +108,107 @@ public final class TemporaryFile {
...
-    /// sure the returning path has a trailing forward slash.
+    /// Creates a temporary directory which is automatically removed when the object of this class goes out of scope.
     ///
     /// - Parameters:
-    ///     - dir: If present this will be the temporary directory.
+    ///     - dir: If specified the temporary directory will be created in this directory otherwise enviornment variables
+    ///            TMPDIR, TEMP and TMP will be checked for a value (in that order). If none of the env variables are
+    ///            set, dir will be set to `/tmp/`.
+    ///     - prefix: The prefix to the temporary file name.
+    ///     - removeTreeOnDeinit: If enabled try to delete the whole directory tree even if its not empty.
```
Comment (ddunbar) [inline Sources/Basic/TemporaryFile.swift]: This should document the behavior that the directory is always attempted to be removed, it is just whether that removal is recursive or not that is controlled by the flag.

Comment (aciidb0mb3r) [inline Sources/Basic/TemporaryFile.swift]: That is documented in the main brief above.

**Comment by ddunbar** [comment]
LGTM, I added a couple minor comments on the API and docs.

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by aciidb0mb3r** [comment]
Made changes, will push once CI returns

**Comment by aciidb0mb3r** [comment]
@swift-ci please test

**Pull request merged**